File: csp/stm32l476/swift/Stm32l476UsartBinding.swift
// HAND-WRITTEN Swift descriptor for the STM32L476 USART driver -- the value bundle one bound
// USART wiring resolves to. Every field arrives as a generated literal from a board's
// *Bindings enum; the driver never holds a base address, a pin register, or a divisor of its
// own.
//
// The pin-mode and alternate-function words are per-PORT rather than per-pin: this family's
// USART binding routes its TX and RX pins through ONE MODER read-modify-write and ONE AFRL
// read-modify-write, because a bound pair sits in the same halves of both registers. That is
// why a single reg/mask/value triple describes each, and it is the generator's resolution --
// not an assumption this file is free to make.

/// The descriptor a STM32L476 USART driver consumes: one binding's resolved values, exactly
/// the constants a board's generated *Bindings enum carries. A board class constructs it from
/// those generated literals.
public struct Stm32l476UsartBinding {
    /// The bound USART instance's base address.
    public let base: UInt32
    /// The RCC register gating the USART instance's clock, and its enable bit as a mask.
    public let rccEnReg: UInt32
    public let rccEnMask: UInt32
    /// The RCC register gating the GPIO PORT the pins live on, and its enable bit as a mask.
    /// A separate bank from the USART's on this family -- that split is a generated fact.
    public let portRccEnReg: UInt32
    public let portRccEnMask: UInt32
    /// The port's pin-mode register, the mask covering the bound pins, and the alternate-function
    /// mode value for them.
    public let moderReg: UInt32
    public let moderMask: UInt32
    public let moderValue: UInt32
    /// The port's alternate-function-low register, the mask covering the bound pins, and the
    /// alternate-function selection for them.
    public let afrlReg: UInt32
    public let afrlMask: UInt32
    public let afrlValue: UInt32
    /// The BRR divisor for the binding's wire rate under the board's default clock plan
    /// (plan-derived; never authored).
    public let brrDivisor: UInt32

    public init(base: UInt32, rccEnReg: UInt32, rccEnMask: UInt32,
                portRccEnReg: UInt32, portRccEnMask: UInt32,
                moderReg: UInt32, moderMask: UInt32, moderValue: UInt32,
                afrlReg: UInt32, afrlMask: UInt32, afrlValue: UInt32,
                brrDivisor: UInt32) {
        self.base = base
        self.rccEnReg = rccEnReg
        self.rccEnMask = rccEnMask
        self.portRccEnReg = portRccEnReg
        self.portRccEnMask = portRccEnMask
        self.moderReg = moderReg
        self.moderMask = moderMask
        self.moderValue = moderValue
        self.afrlReg = afrlReg
        self.afrlMask = afrlMask
        self.afrlValue = afrlValue
        self.brrDivisor = brrDivisor
    }
}

File: csp/stm32l476/csharp/Stm32l476UsartBinding.cs
// The descriptor an STM32L476 USART driver consumes: one binding's resolved values, exactly the
namespace Lamella.Boards
{
    /// <summary>The resolved mux facts for ONE bound pin: the port clock gate to open, and the
    /// two read-modify-writes that put the pin into alternate-function mode on its function
    /// number.</summary>
    public sealed class Stm32l476UsartPinMux
    {
        /// <summary>The RCC register gating this pin's port clock.</summary>
        public readonly uint PortRccEnReg;
        /// <summary>That port's clock-enable bit, as a mask.</summary>
        public readonly uint PortRccEnMask;
        /// <summary>The GPIO MODER register covering this pin.</summary>
        public readonly uint ModerReg;
        /// <summary>This pin's MODER field span, to clear before setting.</summary>
        public readonly uint ModerMask;
        /// <summary>The MODER value selecting alternate-function mode for this pin.</summary>
        public readonly uint ModerValue;
        /// <summary>The GPIO alternate-function register covering this pin -- AFRL for pins 0..7,
        /// AFRH for pins 8..15. Which one it is has already been decided.</summary>
        public readonly uint AfrReg;
        /// <summary>This pin's alternate-function nibble, to clear before setting.</summary>
        public readonly uint AfrMask;
        /// <summary>The alternate-function selection for this pin (a DATASHEET fact -- the AF
        /// number appears nowhere in the reference manual).</summary>
        public readonly uint AfrValue;

        public Stm32l476UsartPinMux(uint portRccEnReg, uint portRccEnMask, uint moderReg,
            uint moderMask, uint moderValue, uint afrReg, uint afrMask, uint afrValue)
        {
            PortRccEnReg = portRccEnReg;
            PortRccEnMask = portRccEnMask;
            ModerReg = moderReg;
            ModerMask = moderMask;
            ModerValue = moderValue;
            AfrReg = afrReg;
            AfrMask = afrMask;
            AfrValue = afrValue;
        }
    }

    public sealed class Stm32l476UsartBinding
    {
        /// <summary>The bound USART instance's base address.</summary>
        public readonly uint UsartBase;
        /// <summary>The RCC register gating the USART instance's clock.</summary>
        public readonly uint RccEnReg;
        /// <summary>The USART instance's clock-enable bit, as a mask.</summary>
        public readonly uint RccEnMask;
        /// <summary>The transmit pin's mux facts.</summary>
        public readonly Stm32l476UsartPinMux Tx;
        /// <summary>The receive pin's mux facts.</summary>
        public readonly Stm32l476UsartPinMux Rx;
        /// <summary>The BRR divisor for the binding's wire rate under the board's default clock
        /// plan (plan-derived, so it names its clock source at the board stratum).</summary>
        public readonly uint BaudDivisor;

        public Stm32l476UsartBinding(uint usartBase, uint rccEnReg, uint rccEnMask,
            Stm32l476UsartPinMux tx, Stm32l476UsartPinMux rx, uint baudDivisor)
        {
            UsartBase = usartBase;
            RccEnReg = rccEnReg;
            RccEnMask = rccEnMask;
            Tx = tx;
            Rx = rx;
            BaudDivisor = baudDivisor;
        }
    }
}
